### PR TITLE
MOSIP-27361, MOSIP-27376

### PR DESCRIPTION
**Pull request by HimajaDhanyamraju2**

### CODE REVIEW

File: registration/registration-client/src/main/java/io/mosip/registration/controller/auth/LoginController.java [hunk 1]
```diff
@@ -387,14 +407,45 @@ private void executeSQLFile() {
...
+				Manifest manifest = new Manifest(new FileInputStream(localManifestFile));
+				String backupVersion = manifest.getMainAttributes().getValue(Attributes.Name.MANIFEST_VERSION);
+				/*
+				 * If the backup path is having only one folder inside, we directly consider the
+				 * version inside Manifest file as the previous version. If the path has
+				 * multiple files, we assume that the entry in version-mappings list with
+				 * releaseOrder as "1" will be among the backup files and that is the previous
+				 * version.
+				 */
+				if (file.listFiles().length == 1 || backupVersion.equals(versionMappings.entrySet().iterator().next().getKey())) {
```
Comment (ase-101) [inline registration/registration-client/src/main/java/io/mosip/registration/controller/auth/LoginController.java:439]: This check should be before for loop @HimajaDhanyamraju2

File: registration/registration-services/src/main/java/io/mosip/registration/constants/RegistrationConstants.java [hunk 2]
```diff
@@ -1493,6 +1493,9 @@ public static List<String> fieldsToExclude() {
 	public static final String PUBLIC_KEY_REF_ID = "packet-encryption-key";
 	public static final String USER_DETAIL_SALT_SERVICE_NAME = "user_salt_details";
 	public static final String SERVICES_VERSION_KEY = "mosip.reg.services.current.version";
+	public static final String VERSION_MAPPINGS_KEY = "mosip.registration.verion.upgrade.version-mappings";
+	public static final String ROLLBACK_FILE_PATH = "mosip.reg.rollback.path";
+	public static final String MANIFEST_PATH = "/MANIFEST.MF";
```
Comment (ase-101) [inline registration/registration-services/src/main/java/io/mosip/registration/constants/RegistrationConstants.java:1498]: lets not hardcode the "/", instead when creating the full path user File.separator, Also cross-check how are we taking "mosip.reg.rollback.path"

File: registration/registration-services/src/main/java/io/mosip/registration/dto/VersionMappings.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package io.mosip.registration.dto;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class VersionMappings {
```
Comment (ase-101) [inline registration/registration-services/src/main/java/io/mosip/registration/dto/VersionMappings.java:10]: Please add JsonIgnoreOnUnknown annotation to this DTO

File: registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java [hunk 4]
```diff
@@ -548,96 +551,97 @@ public Timestamp getLatestVersionReleaseTimestamp() {
...
 	 * @throws IOException
 	 */
-	public ResponseDTO executeSqlFile(String actualLatestVersion, String previousVersion) throws IOException {
-
-		LOGGER.info("DB-Script files execution started from previous version : {} , To Current Version : {}",previousVersion, currentVersion);
-
-		String newVersion = actualLatestVersion.split("-")[0];
-		previousVersion = previousVersion.split("-")[0];
+	public ResponseDTO executeSqlFile(String previousVersion, Map<String, VersionMappings> versionMappings) throws IOException {
+		LOGGER.info("DB-Script files execution started from previous version : {} , To Current Version : {}", previousVersion, currentVersion);
```
Comment (ase-101) [inline registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java:579]: where is the "currentVersion" is declared ?

File: registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java [hunk 5]
```diff
@@ -548,96 +551,97 @@ public Timestamp getLatestVersionReleaseTimestamp() {
...
+		
+		/*
+		 * Here, we are removing the entries from version-mappings map, for which, the
+		 * releaseOrder is less than or equal to the previous version, because, we need
+		 * to execute the upgrade scripts only for the versions released after the
+		 * previous version.
+		 */
+		if (versionMappings.containsKey(previousVersion)) {
+			Integer previousVersionReleaseOrder = versionMappings.get(previousVersion).getReleaseOrder();
+			versionMappings.entrySet().removeIf(matches -> matches.getValue().getReleaseOrder() <= previousVersionReleaseOrder);
```
Comment (ase-101) [inline registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java:591]: rename "matches" to versionMapping

File: registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java [hunk 6]
```diff
@@ -548,96 +551,97 @@ public Timestamp getLatestVersionReleaseTimestamp() {
...
-
 				LOGGER.error(LoggerConstants.LOG_REG_UPDATE, APPLICATION_NAME, APPLICATION_ID,
 						exception.getMessage() + ExceptionUtils.getStackTrace(exception));
-
 			}
-			// Prepare Error Response
-			setErrorResponse(responseDTO, RegistrationConstants.SQL_EXECUTION_FAILURE, null);
-
-			// Replace with backup
-			rollback(responseDTO, previousVersion);
```
Comment (ase-101) [inline registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java:624]: Are we handling the backup restore in someother place?

File: registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java [hunk 7]
```diff
@@ -548,96 +551,97 @@ public Timestamp getLatestVersionReleaseTimestamp() {
...
 
+	private void executeSQL(String dbVersion) throws RegBaseCheckedException {
 		try {
-
 			LOGGER.info(LoggerConstants.LOG_REG_UPDATE, APPLICATION_NAME, APPLICATION_ID,
-					"Checking Started : " + newVersion + SLASH + exectionSqlFile);
+					"Checking Started : " + dbVersion + SLASH + exectionSqlFile);
 
-			execute(SQL + SLASH + newVersion + SLASH + exectionSqlFile);
+			execute(SQL + SLASH + dbVersion + SLASH + exectionSqlFile);
```
Comment (ase-101) [inline registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java:622]: After the successful execution of the script, we need to take a backup of migrated DB

File: registration/registration-services/src/main/java/io/mosip/registration/dto/VersionMappings.java (hunk 3 above)
Comment (HimajaDhanyamraju2) [reply registration/registration-services/src/main/java/io/mosip/registration/dto/VersionMappings.java:10]: Done.

File: registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java (hunk 5 above)
Comment (HimajaDhanyamraju2) [reply registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java:591]: Done.

File: registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java (hunk 6 above)
Comment (HimajaDhanyamraju2) [reply registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java:624]: No, we only have it at this place.

File: registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java (hunk 7 above)
Comment (HimajaDhanyamraju2) [reply registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateHandler.java:622]: Yes, added logic to take backup after every successful execution.

**Review by ase-101 — APPROVED** [review]

**Pull request merged**